### PR TITLE
Add support for .ar archives (and .deb files)

### PR DESCRIPTION
**Pull request by kjlubick**
This implements #[15130](https://github.com/bazelbuild/bazel/issues/15130).

As I was updating the docs for .ar and .deb formats, I also addressed some previous formats that had been added but not propagated through to all the documentation places.

### CODE REVIEW

**Comment by aiuto** [comment]
I would like to find a way do to this without adding more code in Bazel. Perhaps we can do this mostly in starlark defined workspace rules.  I'm very familiar with the .deb and .ar formats on the writing side, so I'll take a look at this, but it might not be until next week.

**Comment by meteorcloudy** [comment]
>  Perhaps we can do this mostly in starlark defined workspace rules.

How do you plan to do that? I think this PR extends the functionality of `repository_ctx.download_and_extract`, which is implemented in native code. Not sure how it could be implemented in pure starlark.

**Comment by kjlubick** [comment]
I agree with @meteorcloudy, to implement this in pure starlark would involve decoding .ar files in Starlark code. This is not impossible, but certainly difficult given that we wouldn't really be able to use libraries. Another option is to shell out to a native executable, but that requires users to install or download such a binary.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/ArFunction.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                    continue;
+                } else {
+                    // We do not have to worry about symlinks in .ar files - it's not supported
+                    // by the .ar file format.
+                    try (OutputStream out = filePath.getOutputStream()) {
+                        ByteStreams.copy(arStream, out);
+                    }
+                    filePath.chmod(entry.getMode());
+
+                    Date lastModified = entry.getLastModifiedDate();
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/ArFunction.java:68]: Getting this error message from internal presubmit check:
"Date has a bad API that leads to bugs; prefer java.time.Instant or LocalDate."

File: src/main/java/com/google/devtools/build/lib/bazel/repository/ArFunction.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.bazel.repository;
+
+import com.google.common.io.ByteStreams;
+import com.google.devtools.build.lib.bazel.repository.DecompressorValue.Decompressor;
+import com.google.devtools.build.lib.vfs.Path;
+
+import java.io.*;
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/ArFunction.java:21]: Wildcard imports, static or otherwise, should not be used

**Review by meteorcloudy — CHANGES_REQUESTED** [review]

**Comment by ckolli5** [comment]
@bazel-io fork 5.2.0

**Comment by ckolli5** [comment]
@bazel-io fork 5.2.0

**Comment by uhthomas** [comment]
Fixes https://github.com/bazelbuild/bazel/issues/13905?

**Pull request closed without merging**